File: week-01/p4_advance.go
package main

import "fmt"

// Advanced solution using a 2-D Prefix Sum (Summed Area Table).
//
// ------------------
// Naive approach (sumRange in p4.go):
//   - Each range query loops over every cell in the sub-matrix.
//   - Time per query: O(n * m) --> O(n^2) <-- bad when many queries are made
//
// Prefix-sum approach (this file):
//   - One-time O(R * C) build step to pre-compute a prefix-sum table.
//   - Every range query after that is answered in O(1) with 4 table
//     look-ups and 3 arithmetic operations.
//   - Time per query: O(1) --> O(1) <-- constant, regardless of region size
//
// Total cost for Q queries:
//   Naive       : O(Q * n * m)
//   Prefix-sum  : O(R*C) + O(Q)   ≈ O(R*C + Q)

// buildPrefixSum builds a (rows+1) x (cols+1) prefix-sum table using
// 1-based indexing so edge cases (row 0 / col 0) need no bounds checks.
//
// ps[i][j] = sum of all cells matrix[r][c] where 0 <= r < i and 0 <= c < j
//
// Recurrence:
//
//	ps[i][j] = ps[i-1][j] + ps[i][j-1] - ps[i-1][j-1] + matrix[i-1][j-1]
//
// Build time : O(R * C)
// Space      : O(R * C)
func buildPrefixSum(matrix [][]int) [][]int {
	rows := len(matrix)
	cols := len(matrix[0])

	// Allocate (rows+1) x (cols+1) — the extra row/col of zeros is the sentinel border.
	ps := make([][]int, rows+1)
	for i := range ps {
		ps[i] = make([]int, cols+1)
	}

	for i := 1; i <= rows; i++ {
		for j := 1; j <= cols; j++ {
			ps[i][j] = ps[i-1][j] +
				ps[i][j-1] -
				ps[i-1][j-1] +
				matrix[i-1][j-1]
		}
	}
	return ps
}

// efficientSumRange returns the sum of the sub-matrix defined by topLeft
// and bottomRight (0-based, inclusive on both ends) using the pre-built
// prefix-sum table.
func efficientSumRange(ps [][]int, topLeft, bottomRight [2]int) int {
	r1, c1 := topLeft[0], topLeft[1]
	r2, c2 := bottomRight[0], bottomRight[1]
	return ps[r2+1][c2+1] -
		ps[r1][c2+1] -
		ps[r2+1][c1] +
		ps[r1][c1]
}

func main() {
	matrix := [][]int{
		{1, 2, 3, 4, 5},
		{8, 6, 9, 1, 3},
		{8, 3, 1, 4, 3},
		{4, 8, 2, 9, 6},
	}
	ps := buildPrefixSum(matrix)
	fmt.Println(efficientSumRange(ps, [2]int{1, 2}, [2]int{2, 3}))
}

File: practices/materials/in-class/week-01/p4_advance.py
"""
Advanced solution using a 2-D Prefix Sum (Summed Area Table).

------------------
Naive approach (p4.py  sum_range):
  - Each range query loops over every cell in the sub-matrix.
  - Time per query: O(n * m) --> O(n^2) <-- bad when many queries are made

Prefix-sum approach (this file):
  - One-time  O(R * C)  build step to pre-compute prefix_sum[][].
  - Every range query after that is answered in O(1) with 4 table
    look-ups and 3 arithmetic operations.
  - Time per query: O(1) <-- constant, regardless of region size

This matters when the same matrix is queried many times (Q queries):
  Naive        : O(Q * n * m)
  Prefix-sum   : O(R*C)  +  O(Q)   ≈ O(R*C + Q)
"""


def build_prefix_sum(matrix: list[list[int]]) -> list[list[int]]:
    """
    Build a (rows+1) x (cols+1) prefix-sum table using 1-based indexing
    so that edge cases (row 0 / col 0) never require bounds checks.

    prefix_sum[i][j] = sum of all cells matrix[r][c]
                       where 0 <= r < i  and  0 <= c < j

    Recurrence:
        prefix_sum[i][j] = prefix_sum[i-1][j]
                         + prefix_sum[i][j-1]
                         - prefix_sum[i-1][j-1]   # subtracted twice above
                         + matrix[i-1][j-1]

    Build time : O(R * C)
    Space      : O(R * C)
    """
    rows = len(matrix)
    cols = len(matrix[0])
    # Extra row and column of zeros act as a sentinel border
    ps = [[0] * (cols + 1) for _ in range(rows + 1)]

    for i in range(1, rows + 1):
        for j in range(1, cols + 1):
            ps[i][j] = (
                ps[i - 1][j]
                + ps[i][j - 1]
                - ps[i - 1][j - 1]
                + matrix[i - 1][j - 1]
            )
    return ps


def efficient_sum_range(
    prefix_sum: list[list[int]],
    top_left: tuple[int, int],
    bottom_right: tuple[int, int],
) -> int:
    """
    Return the sum of the sub-matrix defined by top_left and bottom_right
    (0-based, inclusive on both ends) using the pre-built prefix-sum table.

    Formula (inclusion-exclusion on the Summed Area Table):

        sum = ps[r2+1][c2+1]
            - ps[r1  ][c2+1]   # strip above the region
            - ps[r2+1][c1  ]   # strip left of the region
            + ps[r1  ][c1  ]   # re-add the top-left corner (subtracted twice)

    Query time: O(1)  <-- only 4 table lookups + 3 additions
    """
    r1, c1 = top_left
    r2, c2 = bottom_right
    return (
        prefix_sum[r2 + 1][c2 + 1]
        - prefix_sum[r1][c2 + 1]
        - prefix_sum[r2 + 1][c1]
        + prefix_sum[r1][c1]
    )


if __name__ == "__main__":
    matrix = [[1, 2, 3, 4, 5], [8, 6, 9, 1, 3], [8, 3, 1, 4, 3], [4, 8, 2, 9, 6]]

    ps = build_prefix_sum(matrix)

    result = efficient_sum_range(ps, (1, 2), (2, 3))
    print(f"efficient_sum_range (1,2)->(2,3) = {result}")

    # Multiple queries all run in O(1) after the single O(R*C) build
    queries = [
        ((0, 0), (3, 4)),   # entire matrix
        ((0, 0), (0, 4)),   # first row
        ((0, 0), (3, 0)),   # first column
        ((1, 1), (2, 2)),   # inner 2x2 block
    ]
    for tl, br in queries:
        print(f"  sum {tl} -> {br} = {efficient_sum_range(ps, tl, br)}")
